### PR TITLE
4712-Text-Core-package-documentation-and-test

### PR DESCRIPTION
**Pull request by hilaire**
WIP #4712
On TextStream

### CODE REVIEW

**Comment by Ducasse** [comment]
This is cool!

**Comment by MarcusDenker** [comment]
should we merge this already?

File: src/Text-Tests/TextStreamTest.class.st [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+{ #category : #test }
+TextStreamTest >> testNextPutAll [
+	stream 
+		nextPutAll: 'Pharo';
+		space.
+	self assert: stream contents isText.
+	self assert: stream contents string equals: 'Pharo '.
+	"No attribute was set"
+	1 to: 6 do: [:index | self assert: (stream contents runs at: index) isEmpty]
```
Comment (jecisc) [inline src/Text-Tests/TextStreamTest.class.st]:
```suggestion
	1 to: 6 do: [:index | self assertEmpty: (stream contents runs at: index) ]
```

File: src/Text-Tests/TextStreamTest.class.st [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	stream nextPutAll: ' is cool'.
+	self assert: stream contents string equals: 'Pharo is cool'.
+	1 to: 5 do: [ :index  |
+		attributes := stream contents runs at: index.
+		self assert: attributes size equals: 2.
+		self assert: attributes first emphasisCode equals:  1.
+		self assert: attributes second color equals: Color yellow ].
+	"No attribute for the remaining part of the text"
+	6 to: 12 do: [ :index |
+		self assert: (stream contents runs at: index ) isEmpty]
```
Comment (jecisc) [inline src/Text-Tests/TextStreamTest.class.st]:
```suggestion
		self assertEmpty: (stream contents runs at: index ) ]
```

File: src/Text-Tests/TextStreamTest.class.st [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		withAttribute: TextEmphasis underlined 
+		do: [stream nextPutAll: 'is cool'  ].
+	self assert: stream contents string equals: 'Pharo is cool'.
+	"part of the text underlined"
+	1 to: 5 do: [ :index |
+		attributes := stream contents runs at: index.
+		self assert: attributes isArray.
+		self assert: attributes first emphasisCode equals: 1].
+	"no attribute on this space character"
+	self assert: (stream contents runs at: 6) isEmpty.
```
Comment (jecisc) [inline src/Text-Tests/TextStreamTest.class.st]:
```suggestion
	self assertEmpty: (stream contents runs at: 6).
```

**Review by jecisc — APPROVED** [review]
Cool!

I suggested some changes to have better logs in the CI in case it fails in the future.

**Comment by hilaire** [comment]
@MarcusDenker About merging, I don't know but I have not finished. Is it ok after merge to add commit to this PR or should I open another PR?

**Comment by jecisc** [comment]
You can still add commits on the branch, but you’ll need to open a new PR

**Comment by hilaire** [comment]
@jecisc, I fetched these 3 tiny changes and Iceberg is pulling a full set of changes not related. I specifically asked a pull for these 3 changes. Took too long and I have now unwished changes in my image I will have to pay attention on when commiting.

Btw, I can't create a new PR on the same branch. It is refused. Or do I miss something?

Capture

**Comment by hilaire** [comment]
@jecisc , looks like the commit got it way to the existing PR even if merged. But was it merged after all?

**Comment by hilaire** [comment]
See the comment of TextAttribute. What is still relevant in this comment

```
Tells a piece of text to be a certain way.

Select text, press Command-6, choose a attribute.  If selected text is of the form 
	Hi There<Smalltalk beep>
the part in angle brackets is saved for action, and the Hi There appears in the paragraph.  If selection has no angle brackets, use the whole thing as both the text and the action.

TextDoIt  --  eval as a Smalltalk expression (the part in angle brackets)

TextLink -- Show a method, class comment, class hierarchy, or class defintion.
	<Point extent:>, <Point Comment>, <Point Hierarchy>, or <Point Defintion> are what you type.

TextURL -- Show the web page. <www.disney.com>

These attributes of text need to be stored on the disk in a regular file-out.  It is done in this form: 	Hi There   
	in the text, and a Run containing   dSmalltalk beep;;
	Click here to see the extent:   
	in the text, and a Run containing   method LPoint extent:;
See RunArray class scanFrom: where decoding is done.

```

**Comment by hilaire** [comment]
I propose this new comment for TextAttribute. Once I learn enough of TextAttribute I can amend it.
```
I am an abstract class to represent a text attribute, the way a portion of text is graphically represented.
My sub-classes specify these representations:
- colouring (TextColor), 
- bold, italic, underlined, 
- narrow, strike out (TextEmphasis), 
- alignment (TextAlignment),
- indentation (TextIndent),
- font change (TextFontChange),
- kerning (TextKern),
- url and related action (TextAction hierarchy).

My instances are stored in the Text instance runs variable (a RunArray).
```

**Comment by hilaire** [comment]
I see the PluggableTextAttribute with no user. Do we need it? Can we deprecate it?

**Comment by MarcusDenker** [comment]
Should this be merged?

I propose to do small PRs that we can merge fast...  especially tests and comments

**Comment by hilaire** [comment]
Yes, you can merge. I will come back to it later then.

**Comment by MarcusDenker** [comment]
Two tests are failing:

testNoUnusedTemporaryVariablesLeft
the following methods have unused temporary variables and should be cleaned: an OrderedCollection(TextEmphasisTest>>#testDominates)

SystemDependenciesTest>>testExternalMorphicDependencies
Given Collections do not match! additions : #(#'WebBrowser-Core') missing: #()

**Comment by hilaire** [comment]
Fix the first failing test. The second fail, AFAIK, has nothing to do with my commits in this pull.

**Comment by Ducasse** [comment]
Hilaire I do not get why when I look at the files changed I have  
HashAndEqualsTestCase >> testEquality [
 	"Check that TextFontChanges report equality correctly"
 	"Check TextAttribute subclasses with equality message report it correctly"

????

**Comment by Ducasse** [comment]
also SycAbstractAllInstVarAccessorsCommand class

**Comment by hilaire** [comment]
Ah this is unwished comits. Can you delete these changes?

**Comment by Ducasse** [comment]
No we cannot (for now and the soon future) cherry pick on merge.

**Comment by hilaire** [comment]
Fix it. 
Is there any left unwished changes?

**Comment by Ducasse** [comment]
Thanks Hilaire. 
Let's me check.

**Comment by Ducasse** [comment]
Did you remove this change

 @@ -21,7 +21,7 @@ HashAndEqualsTestCase >> setUp [

 { #category : #tests }
 HashAndEqualsTestCase >> testEquality [
 	"Check TextAttribute subclasses with equality message report it correctly"
 	"Check that TextAttribute subclasses report equality correctly"
 	prototypes
 		do: [:p | self
 				should: [(EqualityTester with: p) result]] 

because it looks still there.

**Comment by hilaire** [comment]
That's correct, I updated the comment.

**Comment by Ducasse** [comment]
Ok

**Comment by MarcusDenker** [comment]
nevertheless, now in the build we have two tests failing that did not fail before:

https://github.com/pharo-project/pharo/issues/4818
https://github.com/pharo-project/pharo/issues/4819

I checked and for the SystemDependencies, it is indeed this PR that makes them faill

File: src/SystemCommands-VariableCommands/SycAbstractAllInstVarAccessorsCommand.class.st [hunk 4]
```diff
@@ -1,66 +0,0 @@
-"
-I command to launch the Abstract refactorings: create accessors and abstract the direct instance variable uses into accessors.
```
Comment (jecisc) [inline src/SystemCommands-VariableCommands/SycAbstractAllInstVarAccessorsCommand.class.st:2]: Was this class removal wanted?

Comment (hilaire) [reply src/SystemCommands-VariableCommands/SycAbstractAllInstVarAccessorsCommand.class.st:2]: I don't understand this change. At some point, Iceberg did some unwished changes to my working image. I tried to revert it. Was it a left over, I can't tell for sure.

**Comment by tesonep** [comment]
We have to remove the dependency of WebBrowser from Text, it does not have any sense there. 
It should be handled by the block in the text element. 
If not, we will revert the whole PR

**Comment by Ducasse** [comment]
Yes this is strange since I did not see the dependencies when I checked the tests. 
@Cyril probably hilaire wanted to remove a change from within his PR and it was converted to a removal. 
I should say that I do not really know how to use iceberg for cherry picking.

**Comment by Ducasse** [comment]
Pablo technically can hilaire still do it in his PR or should we create a new issue to fix it.

**Comment by hilaire** [comment]
TextUrl attribute needs to know about WebBrowser to open an URL, so there is sense.
Will adding a "Smalltalk tools webBrowser" solve the dependency?

**Comment by hilaire** [comment]
@tesonep, what about my previous question?

> TextUrl attribute needs to know about WebBrowser to open an URL, so there is sense.
> Will adding a "Smalltalk tools webBrowser" solve the dependency?

If so I can open an issue.

**Pull request merged**